File: JavaRewrite/src/main/java/io/github/deadshot465/assignments/musicians/Musician.java
package io.github.deadshot465.assignments.musicians;

import java.util.Random;

public class Musician {
    public static final int delay = 750;

    private final String name;
    private final String role;
    private final SkillLevel skillLevel;
    private boolean isFired;

    private static final String[] firstNames = new String[] {
            "Valerie", "Arnold", "Carlos", "Dorothy", "Keesha", "Phoebe", "Ralphie", "Tim", "Wanda",
            "Janet", "Leo", "Yuhei", "Carson"
    };

    private static final String[] lastNames = new String[]{
            "Frizzle",
            "Perlstein",
            "Ramon",
            "Ann",
            "Franklin",
            "Terese",
            "Tennelli",
            "Jamal",
            "Li",
            "Perlstein",
            "Fujioka",
            "Ito",
            "Hage"
    };

    public Musician(String role, SkillLevel skillLevel) {
        this.name = pickName();
        this.role = role;
        this.skillLevel = skillLevel;
        this.isFired = false;

        System.out.printf("Musician %s, playing the %s entered the room.\n", name, role);
    }

    public void setFired(boolean fired) {
        isFired = fired;
    }

    public boolean playSound() {
        if (isFired) {
            System.out.printf("%s just got back to playing in the subway.\n", name);
            return false;
        }

        switch (skillLevel) {
            case Good -> {
                System.out.printf("%s produced sound!\n", name);
                return true;
            }
            case Bad -> {
                final var rng = new Random();
                final var failed = rng.nextInt(0, 5) == 0;
                if (failed) {
                    System.out.printf("%s played a false note. Uh oh.\n", name);
                    System.out.printf("%s sucks! kicked that member out of the band! (%s)\n", name, role);
                } else {
                    System.out.printf("%s produced sound!\n", name);
                }

                return !failed;
            }
            default -> {
                return true;
            }
        }
    }

    private static String pickName() {
        final var rng = new Random();
        final var firstName = firstNames[rng.nextInt(0, firstNames.length)];
        final var lastName = lastNames[rng.nextInt(0, lastNames.length)];
        return firstName + " " + lastName;
    }
}

File: KotlinRewrite/src/main/kotlin/io/github/deadshot465/assignments/musicians/Musician.kt
package io.github.deadshot465.assignments.musicians

import kotlinx.coroutines.channels.Channel
import kotlin.random.Random

const val delay: Long = 750

private val firstNames = listOf("Valerie", "Arnold", "Carlos", "Dorothy", "Keesha", "Phoebe", "Ralphie", "Tim", "Wanda",
    "Janet", "Leo", "Yuhei", "Carson")

private val lastNames = listOf("Frizzle",
    "Perlstein",
    "Ramon",
    "Ann",
    "Franklin",
    "Terese",
    "Tennelli",
    "Jamal",
    "Li",
    "Perlstein",
    "Fujioka",
    "Ito",
    "Hage")

private fun pickName(): String {
    val firstName = firstNames[Random.nextInt(0, firstNames.size)]
    val lastName = lastNames[Random.nextInt(0, lastNames.size)]
    return "$firstName $lastName"
}

data class Musician(val role: String, val skillLevel: SkillLevel, val receiver: Channel<Int>) {
    private val name: String = pickName()

    init {
        println("Musician ${name}, playing the $role entered the room.")
    }

    fun playSound(): Boolean {
        val receiveResult = receiver.tryReceive()
        if (receiveResult.isSuccess) {
            println("$name just got back to playing in the subway.")
            return false
        }

        return when (skillLevel) {
            SkillLevel.Good -> {
                println("$name produced sound!")
                true
            }
            SkillLevel.Bad -> {
                val failed = Random.nextInt(0, 5) == 0
                if (failed) {
                    println("$name played a false note. Uh oh.")
                    println("$name sucks! kicked that member out of the band! ($role)")
                } else {
                    println("$name produced sound!")
                }
                !failed
            }
        }
    }
}
